Optimise ring.DoBatch()

bboreham opened the pull request:
`ring.DoBatch` contributes 20% of the memory allocation in my production distributors.

I started looking at this and realised the existing benchmark was not exercising the functionality - the `cas()` didn't update the ring so it was actually empty, and some test fixture code was also quite far up the profile.

Having fixed all that and made the benchmark call the higher-level function `DoBatch()`, the numbers before are:

```
BenchmarkBatch10x100-2     	   10000	    175156 ns/op	   56648 B/op	     334 allocs/op
BenchmarkBatch100x100-2    	    5000	    256146 ns/op	   87675 B/op	     720 allocs/op
BenchmarkBatch100x1000-2   	    1000	   1838020 ns/op	  578726 B/op	    3278 allocs/op
```

after:

```
BenchmarkBatch10x100-2     	   10000	    149385 ns/op	   12061 B/op	      29 allocs/op
BenchmarkBatch100x100-2    	   10000	    208285 ns/op	   30909 B/op	     238 allocs/op
BenchmarkBatch100x1000-2   	    1000	   1558331 ns/op	  112601 B/op	     217 allocs/op
```

The `ReadRing` function signatures have changed a bit, which may impact downstream projects using this code.

tomwilkie commented on pkg/ring/ring.go:
```diff
@@ -25,16 +25,18 @@ const (
 
-	Get(key uint32, op Operation) (ReplicationSet, error)
-	BatchGet(keys []uint32, op Operation) ([]ReplicationSet, error)
+	Get(key uint32, op Operation, descs []IngesterDesc) (ReplicationSet, error)
```
Can you document the new param is there to preserve allocations, and that its contents are not important - if thats true of course...

tomwilkie commented on pkg/ring/ring.go:
```diff
@@ -181,37 +183,16 @@ func migrateRing(desc *Desc) []TokenDesc {
-// for the given keys. The order of the result matches the order of the input.
-func (r *Ring) BatchGet(keys []uint32, op Operation) ([]ReplicationSet, error) {
+// ingesters should be a 0-length slice which can be overwritten (or nil)
+func (r *Ring) Get(key uint32, op Operation, ingesters []IngesterDesc) (ReplicationSet, error) {
```
nit: `ingesters` here, `descs` in interface.

tomwilkie commented on pkg/ring/replication_strategy.go:
```diff
@@ -26,14 +27,15 @@ func (r *Ring) replicationStrategy(ingesters []IngesterDesc, op Operation) (
 			maxFailure--
 		}
 	}
+	liveIngesters = ingesters
```
I would fully support the removal of the use of named return parameters; I find them hard to follow in the review...

tomwilkie commented on pkg/ring/ring.go:
```diff
@@ -25,16 +25,18 @@ const (
 	// ConsulKey is the key under which we store the ring in consul.
 	ConsulKey = "ring"
+
+	maxExpectedReplicationSet = 5 // used to pre-allocate some arrays
```
Can we put this in the file where it is used?

Also, IIUC, this could be called "max replication factor plus one" or something similar?

Finally, sentences for comments pls.

bboreham replied:
The code returns more ingesters than the replication factor under circumstances I don't fully understand. I'm not at all certain that "plus one" is correct otherwise I would have made the constant 4.

tomwilkie commented:
Nice set of optimisations!  Sorry I've been so nit picky.

Once small concern: we're now taking a (read) lock in a loop.  Not super concerned about consistency (the ring might update mid processing a batch) but do you think this will impact CPU usage?

bboreham commented:
> do you think this will impact CPU usage?

The benchmarks show that CPU overall comes down.  Profiling suggests that `sort.Search()` looking for each token is far more CPU-intensive than the lock.

I have addressed review comments; PTAL.
